File: backend/app/schemas/consolidation.py
"""
Pydantic schemas for consolidation and financial statements endpoints.

Request and response models for budget consolidation and financial reporting.
"""

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.configuration import BudgetVersionStatus
from app.models.consolidation import (
    ConsolidationCategory,
    LineType,
    StatementFormat,
    StatementType,
)

# ==============================================================================
# Budget Consolidation Schemas
# ==============================================================================


class ConsolidationLineItemResponse(BaseModel):
    """Schema for individual consolidation line item."""

    id: uuid.UUID
    account_code: str = Field(..., description="PCG account code")
    account_name: str = Field(..., description="Account name")
    consolidation_category: ConsolidationCategory = Field(
        ..., description="Consolidation category"
    )
    is_revenue: bool = Field(..., description="True if revenue, False if expense")
    amount_sar: Decimal = Field(..., description="Amount in SAR")
    source_table: str = Field(..., description="Source planning table")
    source_count: int = Field(..., description="Number of source records aggregated")
    is_calculated: bool = Field(..., description="True if auto-calculated")
    notes: str | None = Field(None, description="Optional notes")
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BudgetConsolidationResponse(BaseModel):
    """Schema for consolidated budget with summary totals."""

    budget_version_id: uuid.UUID
    budget_version_name: str = Field(..., description="Budget version name")
    fiscal_year: int = Field(..., description="Fiscal year")
    academic_year: str = Field(..., description="Academic year")
    status: BudgetVersionStatus = Field(..., description="Budget version status")

    # Line items grouped by type
    revenue_items: list[ConsolidationLineItemResponse] = Field(
        default_factory=list, description="Revenue line items"
    )
    personnel_items: list[ConsolidationLineItemResponse] = Field(
        default_factory=list, description="Personnel cost line items"
    )
    operating_items: list[ConsolidationLineItemResponse] = Field(
        default_factory=list, description="Operating cost line items"
    )
    capex_items: list[ConsolidationLineItemResponse] = Field(
        default_factory=list, description="CapEx line items"
    )

    # Summary totals
    total_revenue: Decimal = Field(..., description="Total revenue")
    total_personnel_costs: Decimal = Field(..., description="Total personnel costs")
    total_operating_costs: Decimal = Field(..., description="Total operating costs")
    total_capex: Decimal = Field(..., description="Total capital expenditure")
    operating_result: Decimal = Field(
        ..., description="Operating result (revenue - personnel - operating)"
    )
    net_result: Decimal = Field(
        ..., description="Net result (operating result - capex if expensed)"
    )

    model_config = ConfigDict(from_attributes=True)


class ConsolidationValidationResponse(BaseModel):
    """Schema for consolidation validation results."""

    is_complete: bool = Field(..., description="Whether all required modules are complete")
    missing_modules: list[str] = Field(
        default_factory=list, description="List of missing module names"
    )
    warnings: list[str] = Field(default_factory=list, description="Validation warnings")
    module_counts: dict[str, int] = Field(
        default_factory=dict, description="Record counts per module"
    )


class ConsolidationRequest(BaseModel):
    """Schema for consolidation request."""

    recalculate: bool = Field(
        default=True, description="Whether to force recalculation"
    )


# ==============================================================================
# Financial Statement Schemas
# ==============================================================================


class FinancialStatementLineResponse(BaseModel):
    """Schema for financial statement line item."""

    id: uuid.UUID
    line_number: int = Field(..., description="Sequential line number for ordering")
    line_type: LineType = Field(..., description="Type of line")
    indent_level: int = Field(..., description="Indentation level (0-3)")
    line_code: str | None = Field(None, description="Account or section code")
    line_description: str = Field(..., description="Line description")
    amount_sar: Decimal | None = Field(None, description="Amount in SAR")
    is_bold: bool = Field(..., description="Display in bold")
    is_underlined: bool = Field(..., description="Display underlined")
    source_consolidation_category: str | None = Field(
        None, description="Source consolidation category"
    )

    model_config = ConfigDict(from_attributes=True)


class IncomeStatementResponse(BaseModel):
    """Schema for income statement."""

    id: uuid.UUID
    budget_version_id: uuid.UUID
    statement_type: StatementType = Field(..., description="Statement type")
    statement_format: StatementFormat = Field(..., description="Format standard")
    statement_name: str = Field(..., description="Statement name")
    fiscal_year: int = Field(..., description="Fiscal year")
    total_amount_sar: Decimal = Field(..., description="Net result")
    is_calculated: bool = Field(..., description="True if auto-calculated")
    notes: str | None = Field(None, description="Optional notes")
    lines: list[FinancialStatementLineResponse] = Field(
        default_factory=list, description="Statement lines"
    )
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BalanceSheetResponse(BaseModel):
    """Schema for balance sheet (assets + liabilities)."""

    budget_version_id: uuid.UUID
    fiscal_year: int = Field(..., description="Fiscal year")
    assets: IncomeStatementResponse = Field(..., description="Assets statement")
    liabilities: IncomeStatementResponse = Field(..., description="Liabilities statement")
    is_balanced: bool = Field(
        ..., description="True if total assets = total liabilities"
    )

    model_config = ConfigDict(from_attributes=True)


class FinancialPeriodTotals(BaseModel):
    """Schema for period-specific financial totals."""

    budget_version_id: uuid.UUID
    period: str = Field(..., description="Period identifier (p1, summer, p2, annual)")
    total_revenue: Decimal = Field(..., description="Total revenue for period")
    total_expenses: Decimal = Field(..., description="Total expenses for period")
    operating_result: Decimal = Field(..., description="Operating result for period")
    net_result: Decimal = Field(..., description="Net result for period")

    model_config = ConfigDict(from_attributes=True)


# ==============================================================================
# Approval Workflow Schemas
# ==============================================================================


class SubmitForApprovalRequest(BaseModel):
    """Schema for submit for approval request."""

    notes: str | None = Field(None, description="Submission notes")


class ApprovebudgetRequest(BaseModel):
    """Schema for approve budget request."""

    notes: str | None = Field(None, description="Approval notes")


class WorkflowActionResponse(BaseModel):
    """Schema for workflow action response."""

    budget_version_id: uuid.UUID
    previous_status: BudgetVersionStatus = Field(..., description="Previous status")
    new_status: BudgetVersionStatus = Field(..., description="New status")
    action_by: uuid.UUID = Field(..., description="User who performed action")
    action_at: datetime = Field(..., description="When action was performed")
    message: str = Field(..., description="Success message")

    model_config = ConfigDict(from_attributes=True)


# ==============================================================================
# Statement Generation Schemas
# ==============================================================================


class GenerateStatementRequest(BaseModel):
    """Schema for statement generation request."""

    format: str = Field(
        default="pcg",
        description="Statement format ('pcg' for French PCG, 'ifrs' for IFRS)",
    )
    regenerate: bool = Field(
        default=False, description="Force regeneration if statement exists"
    )


class StatementFormatOptions(BaseModel):
    """Schema for available statement format options."""

    formats: list[str] = Field(
        default=["pcg", "ifrs"], description="Available format options"
    )
    default_format: str = Field(default="pcg", description="Default format")


# ==============================================================================
# Variance Analysis Schemas
# ==============================================================================


class ConsolidationVarianceItem(BaseModel):
    """Schema for consolidation variance item (comparing two versions)."""

    account_code: str = Field(..., description="PCG account code")
    account_name: str = Field(..., description="Account name")
    category: ConsolidationCategory = Field(..., description="Consolidation category")
    is_revenue: bool = Field(..., description="True if revenue")

    baseline_amount: Decimal = Field(..., description="Baseline version amount")
    comparison_amount: Decimal = Field(..., description="Comparison version amount")
    variance_amount: Decimal = Field(..., description="Absolute variance")
    variance_percent: Decimal = Field(..., description="Percentage variance")

    model_config = ConfigDict(from_attributes=True)


class ConsolidationVarianceResponse(BaseModel):
    """Schema for consolidation variance analysis."""

    baseline_version_id: uuid.UUID
    baseline_version_name: str = Field(..., description="Baseline version name")
    comparison_version_id: uuid.UUID
    comparison_version_name: str = Field(..., description="Comparison version name")

    revenue_variances: list[ConsolidationVarianceItem] = Field(
        default_factory=list, description="Revenue variances"
    )
    expense_variances: list[ConsolidationVarianceItem] = Field(
        default_factory=list, description="Expense variances"
    )

    total_revenue_variance: Decimal = Field(..., description="Total revenue variance")
    total_expense_variance: Decimal = Field(..., description="Total expense variance")
    net_result_variance: Decimal = Field(..., description="Net result variance")

    model_config = ConfigDict(from_attributes=True)


# ==============================================================================
# Summary Schemas
# ==============================================================================


class ConsolidationSummary(BaseModel):
    """Schema for high-level consolidation summary."""

    budget_version_id: uuid.UUID
    budget_version_name: str = Field(..., description="Budget version name")
    fiscal_year: int = Field(..., description="Fiscal year")
    status: BudgetVersionStatus = Field(..., description="Budget version status")

    total_revenue: Decimal = Field(..., description="Total revenue")
    total_expenses: Decimal = Field(..., description="Total expenses (personnel + operating)")
    total_capex: Decimal = Field(..., description="Total capital expenditure")
    operating_result: Decimal = Field(..., description="Operating result")
    net_result: Decimal = Field(..., description="Net result")

    revenue_count: int = Field(..., description="Number of revenue line items")
    expense_count: int = Field(..., description="Number of expense line items")
    capex_count: int = Field(..., description="Number of CapEx line items")

    last_consolidated_at: datetime | None = Field(
        None, description="When last consolidated"
    )

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/services/consolidation.ts
import { apiRequest } from '@/lib/api-client'
import { withServiceErrorHandling } from './utils'

/**
 * Module completion status
 */
export interface ModulesCompleteStatus {
  enrollment: boolean
  classes: boolean
  dhg: boolean
  revenue: boolean
  costs: boolean
  capex: boolean
}

/**
 * Consolidation status response from the backend
 */
export interface ConsolidationStatus {
  version_id: string
  is_complete: boolean
  modules_complete: ModulesCompleteStatus
}

/**
 * Consolidation line item from the backend
 */
export interface ConsolidationLineItem {
  id: string
  account_code: string
  account_name: string
  consolidation_category: string
  is_revenue: boolean
  amount_sar: number
  source_table: string
  source_count: number
  is_calculated: boolean
  notes: string | null
  created_at: string
  updated_at: string
}

/**
 * Full consolidated budget response from the backend
 */
export interface BudgetConsolidation {
  version_id: string
  version_name: string
  fiscal_year: number
  academic_year: string
  status: string
  revenue_items: ConsolidationLineItem[]
  personnel_items: ConsolidationLineItem[]
  operating_items: ConsolidationLineItem[]
  capex_items: ConsolidationLineItem[]
  total_revenue: number
  total_personnel_costs: number
  total_operating_costs: number
  total_capex: number
  operating_result: number
  net_result: number
}

/**
 * Consolidation validation response from the backend
 */
export interface ConsolidationValidation {
  is_complete: boolean
  missing_modules: string[]
  warnings: string[]
  module_counts: Record<string, number>
}

/**
 * Consolidation summary response from the backend
 */
export interface ConsolidationSummary {
  version_id: string
  version_name: string
  fiscal_year: number
  status: string
  total_revenue: number
  total_expenses: number
  total_capex: number
  operating_result: number
  net_result: number
  revenue_count: number
  expense_count: number
  capex_count: number
  last_consolidated_at: string | null
}

/**
 * Workflow action response from the backend (for submit/approve)
 */
export interface WorkflowActionResponse {
  version_id: string
  previous_status: string
  new_status: string
  action_by: string
  action_at: string
  message: string
}

/**
 * Financial statement line item
 */
export interface FinancialStatementLine {
  id: string
  line_number: number
  line_type: string
  indent_level: number
  line_code: string | null
  line_description: string
  amount_sar: number | null
  is_bold: boolean
  is_underlined: boolean
  source_consolidation_category: string | null
}

/**
 * Income statement response
 */
export interface IncomeStatementResponse {
  id: string
  version_id: string
  statement_type: string
  statement_format: string
  statement_name: string
  fiscal_year: number
  total_amount_sar: number
  is_calculated: boolean
  notes: string | null
  lines: FinancialStatementLine[]
  created_at: string
  updated_at: string
}

/**
 * Balance sheet response
 */
export interface BalanceSheetResponse {
  version_id: string
  fiscal_year: number
  assets: IncomeStatementResponse
  liabilities: IncomeStatementResponse
  is_balanced: boolean
}

/**
 * Period financial totals
 */
export interface FinancialPeriodTotals {
  version_id: string
  period: string
  total_revenue: number
  total_expenses: number
  operating_result: number
  net_result: number
}

export const consolidationService = {
  /**
   * Get consolidation status for a version (module completion)
   */
  getStatus: async (versionId: string) => {
    return withServiceErrorHandling(
      apiRequest<ConsolidationStatus>({
        method: 'GET',
        url: `/consolidation/${versionId}/status`,
      }),
      'consolidation: get status'
    )
  },

  /**
   * Get full consolidated budget with all line items
   */
  getConsolidatedBudget: async (versionId: string) => {
    return withServiceErrorHandling(
      apiRequest<BudgetConsolidation>({
        method: 'GET',
        url: `/consolidation/${versionId}`,
      }),
      'consolidation: get consolidated budget'
    )
  },

  /**
   * Trigger budget consolidation (aggregates all planning modules)
   */
  consolidate: async (versionId: string) => {
    return withServiceErrorHandling(
      apiRequest<BudgetConsolidation>({
        method: 'POST',
        url: `/consolidation/${versionId}/consolidate`,
      }),
      'consolidation: run consolidation'
    )
  },

  /**
   * Submit budget for approval (WORKING → SUBMITTED)
   */
  submitForApproval: async (versionId: string, notes?: string) => {
    return withServiceErrorHandling(
      apiRequest<WorkflowActionResponse>({
        method: 'POST',
        url: `/consolidation/${versionId}/submit`,
        data: notes ? { notes } : undefined,
      }),
      'consolidation: submit for approval'
    )
  },

  /**
   * Approve budget (SUBMITTED → APPROVED) - requires manager role
   */
  approve: async (versionId: string, notes?: string) => {
    return withServiceErrorHandling(
      apiRequest<WorkflowActionResponse>({
        method: 'POST',
        url: `/consolidation/${versionId}/approve`,
        data: notes ? { notes } : undefined,
      }),
      'consolidation: approve budget'
    )
  },

  /**
   * Get validation results for budget completeness
   */
  getValidation: async (versionId: string) => {
    return withServiceErrorHandling(
      apiRequest<ConsolidationValidation>({
        method: 'GET',
        url: `/consolidation/${versionId}/validation`,
      }),
      'consolidation: get validation'
    )
  },

  /**
   * Get consolidation summary (totals and counts without line items)
   */
  getSummary: async (versionId: string) => {
    return withServiceErrorHandling(
      apiRequest<ConsolidationSummary>({
        method: 'GET',
        url: `/consolidation/${versionId}/summary`,
      }),
      'consolidation: get summary'
    )
  },

  /**
   * Get income statement
   */
  getIncomeStatement: async (versionId: string, format: 'pcg' | 'ifrs' = 'pcg') => {
    return withServiceErrorHandling(
      apiRequest<IncomeStatementResponse>({
        method: 'GET',
        url: `/consolidation/${versionId}/statements/income`,
        params: { format },
      }),
      'consolidation: get income statement'
    )
  },

  /**
   * Get balance sheet
   */
  getBalanceSheet: async (versionId: string) => {
    return withServiceErrorHandling(
      apiRequest<BalanceSheetResponse>({
        method: 'GET',
        url: `/consolidation/${versionId}/statements/balance`,
      }),
      'consolidation: get balance sheet'
    )
  },

  /**
   * Get financial totals for all periods (P1, Summer, P2, Annual)
   */
  getAllPeriodTotals: async (versionId: string) => {
    return withServiceErrorHandling(
      apiRequest<FinancialPeriodTotals[]>({
        method: 'GET',
        url: `/consolidation/${versionId}/statements/periods`,
      }),
      'consolidation: get all period totals'
    )
  },

  /**
   * Get financial totals for a specific period
   */
  getPeriodTotals: async (versionId: string, period: 'p1' | 'summer' | 'p2' | 'annual') => {
    return withServiceErrorHandling(
      apiRequest<FinancialPeriodTotals>({
        method: 'GET',
        url: `/consolidation/${versionId}/statements/periods/${period}`,
      }),
      'consolidation: get period totals'
    )
  },

  /**
   * Generic financial statement endpoint (for INCOME, BALANCE, or CASHFLOW)
   */
  getStatement: async (
    versionId: string,
    type: 'INCOME' | 'BALANCE' | 'CASHFLOW',
    format: 'PCG' | 'IFRS' = 'PCG',
    period: 'ANNUAL' | 'P1' | 'P2' | 'SUMMER' = 'ANNUAL'
  ) => {
    return withServiceErrorHandling(
      apiRequest<IncomeStatementResponse | BalanceSheetResponse>({
        method: 'GET',
        url: `/consolidation/${versionId}/statements/${type}`,
        params: { format, period },
      }),
      'consolidation: get financial statement'
    )
  },
}
